lib/pry-doc/pry_ext/show_source_with_c_internals/code_fetcher.rb

```diff
@@ -85,61 +88,10 @@ def self.parse_tagfile
 
     def self.tagfile
       tags = File.join(ruby_source_folder, "TAGS")
-      install_and_setup_ruby_source unless File.exists?(tags)
-
-      @tagfile ||= File.read(tags)
-    end
-
-    # @param [String] message Message to display on error
-    # @param [&Block] block Optional assertion
-    def self.check_for_error(message, &block)
-      raise Pry::CommandError, message if $?.to_i != 0 || block && !block.call
-    end
-
-    def self.ask_for_install
-      print "Identifier not found - do you want to install CRuby sources to attempt to resolve the identifier there?\nThis allows the lookup of C internals Y/N "
-
-      if $stdin.gets !~ /^y/i
-        puts "CRuby sources not installed. To prevent being asked again, add `Pry.config.skip_cruby_source = true` to your ~/.pryrc"
-        raise Pry::CommandError, "No definition found."
-      end
-    end
-
-    def self.install_and_setup_ruby_source
-      ask_for_install
-      puts "Downloading and setting up Ruby #{ruby_version} source..."
-      download_ruby
-      generate_tagfile
-      puts "...Finished!"
-    end
-
-    def self.download_ruby
-      curl_cmd = "curl --fail -L https://github.com/ruby/ruby/archive/v#{ruby_version}.tar.gz | tar xzvf - 2> /dev/null"
-
-      FileUtils.mkdir_p(ruby_source_folder)
-      FileUtils.cd(File.dirname(ruby_source_folder)) do
-        %x{ #{curl_cmd} }
-        check_for_error(curl_cmd) { Dir.entries(ruby_source_folder).count > 5 }
-      end
-    end
-
-    def self.etag_binary
-      @etag_binary ||= if RbConfig::CONFIG['host_os'] =~ /linux/
-                         arch = RbConfig::CONFIG['arch'] =~ /i(3|6)86/ ? 32 : 64
-                         File.join(PryDoc.root, "libexec/linux/etags-#{arch}")
-                       else
-                         "etags"
-                       end
-    end
-
-    def self.generate_tagfile
-      find_cmd = "find . -type f -name '*.[chy]' | #{etag_binary} - --no-members"
-
-      FileUtils.cd(ruby_source_folder) do
-        puts "Generating tagfile!"
-        %x{ #{find_cmd} }
-        check_for_error(find_cmd) { File.size(File.join(ruby_source_folder, "TAGS")) > 500 }
+      if !File.exists?(tags)
+        ruby_source_installer.install
       end
+      @tagfile ||= File.read(tags)
```

Comment:
Why do we need `||=` instead of `=`?

Reply:
@kyrylo we don't technically, but might as well memoize as it's an expensive operation, and in the future another method might call it i guess

Reply:
I think we might need to do something similar to what you did in https://github.com/pry/pry-doc/pull/85/commits/97498182940d07ecc213730a6cfa7e33a5332135 because we would be calculating previous lines nevertheless.

Reply:
@kyrylo how so? all the methods are memoized, including callers, an callees

Reply:
It's brittle and if you change the code in the future it's not obvious there's such a requirement for this method. `File.exists?(tags)` isn't memoized.

All we need is to write a guard and change `||=` to `=`:

```rb
def self.tagfile
  return @tagfile if @tagfile
  # ...
  @tagfile = File.read(tags)
end
```

Reply:
I was referring to this comment of yours: "because we would be calculating previous lines nevertheless." which i found pretty confusing :smile: But sure, i can fold `File.exists?` inside